src/dbnode/storage/namespace.go

```diff
@@ -480,19 +488,35 @@ func (n *dbNamespace) AssignShardSet(shardSet sharding.ShardSet) {
 	n.shardSet = shardSet
 	n.shards = make([]databaseShard, n.shardSet.Max()+1)
 	for _, shard := range n.shardSet.AllIDs() {
-		if int(shard) < len(existing) && existing[shard] != nil {
-			n.shards[shard] = existing[shard]
-		} else {
-			bootstrapEnabled := n.nopts.BootstrapEnabled()
+		// We create shards if its an initial assignment or if its not an initial assignment
+		// and the shard doesn't already exist.
+		if initialAssignment ||
+			(int(shard) >= len(existing) || existing[shard] == nil) {
 			n.shards[shard] = newDatabaseShard(metadata, shard, n.blockRetriever,
 				n.namespaceReaderMgr, n.increasingIndex, n.reverseIndex,
-				bootstrapEnabled, n.opts, n.seriesOpts)
-			n.metrics.shards.add.Inc(1)
+				needsBootstrap, n.opts, n.seriesOpts)
+			// NB(bodu): We only record shard add metrics for shards created in non
+			// initial assignments.
+			if !initialAssignment {
+				n.metrics.shards.add.Inc(1)
+			}
+		} else {
+			n.shards[shard] = existing[shard]
 		}
 	}
-	if idx := n.reverseIndex; idx != nil {
-		idx.AssignShardSet(shardSet)
+
+	if !initialAssignment {
```

Comment:
I don't think it would hurt to perform these calls on initial assignment as well. 

Perhaps we don't special case initial assignment for this?

Reply:
(Understand that's what was here before, but probably a good idea to have them all assign the shard set immediately on initialization)